navigator: support generic tail decoration for folder nodes

vince-fugnitto opened the pull request:
<!--
Thank you for your Pull Request. Please provide a description and review
the requirements below.

Contributors guide: https://github.com/theia-ide/theia/blob/master/CONTRIBUTING.md
-->

#### What it does
<!-- Include relevant issues and describe how they are addressed. -->

#### How to test
<!-- Explain how a reviewer can reproduce a bug, test new functionality or verify performance improvements. -->

#### Review checklist

- [ ] as an author, I have thoroughly tested my changes and carefully followed [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#requesting-a-review)

#### Reminder for reviewers

- as a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)

vince-fugnitto commented on packages/core/src/browser/style/tree-decorators.css:
```diff
@@ -79,3 +79,4 @@
     bottom: 40%;
     right: 25%;
 }
+
```
We can revert this file, as the changes are unnecessary.

vince-fugnitto commented on packages/core/src/browser/style/tree.css:
```diff
@@ -189,3 +189,4 @@
     cursor: default;
     opacity: var(--theia-mod-disabled-opacity);
 }
+
```
We can revert this file, as the changes are unnecessary.

vince-fugnitto commented on packages/core/src/browser/tree/tree-widget.tsx:
```diff
@@ -729,54 +729,113 @@ export class TreeWidget extends ReactWidget implements StatefulWidget {
         return icon;
     }
 
+    /**
+     * Render the tree node tail decorations.
+     * @returns the generic class icon used for the tail decoration
+     */
+    protected getGenericClassIconDecoration(): string[] {
```
The documentation is misleading, the method does not do any rendering, it only returns the generic icon decoration.

```suggestion
    /**
     * Get the generic icon decoration.
     * @returns the class icon used for generic decorations.
     */
    protected getGenericClassIconDecoration(): string[] {
```

vince-fugnitto commented on packages/core/src/browser/tree/tree-widget.tsx:
```diff
@@ -729,54 +729,113 @@ export class TreeWidget extends ReactWidget implements StatefulWidget {
+     * @returns the generic class icon used for the tail decoration
+     */
+    protected getGenericClassIconDecoration(): string[] {
+        return ['theia-marker-container-decoration', 'fa', 'fa-circle'];
```
I believe `theia-marker-container-decoration` is incorrect (this is provided by `@theia/markers` which `@theia/core` should not know about).

vince-fugnitto commented on packages/core/src/browser/tree/tree-widget.tsx:
```diff
@@ -729,54 +729,113 @@ export class TreeWidget extends ReactWidget implements StatefulWidget {
+
+        if (isCompositeTreeNode) {
+            // if the node is a composite, we just want to use the decorationData with the highest Priority (last one)
+            let tableauHighPrio: (TreeDecoration.TailDecoration | TreeDecoration.TailDecorationIcon | TreeDecoration.TailDecorationIconClass)[] = [];
```
The naming of the variable is incorrect (we should stick to english as well).

vince-fugnitto commented on packages/core/src/browser/tree/tree-widget.tsx:
```diff
@@ -729,54 +729,113 @@ export class TreeWidget extends ReactWidget implements StatefulWidget {
+        const isCompositeTreeNode = CompositeTreeNode.is(node);
+
+        if (isCompositeTreeNode) {
+            // if the node is a composite, we just want to use the decorationData with the highest Priority (last one)
```
```suggestion
            // If the node is a composite, we just want to use the decorationData with the highest priority (last element).
```

vince-fugnitto commented on packages/core/src/browser/tree/tree-widget.tsx:
```diff
@@ -729,54 +729,113 @@ export class TreeWidget extends ReactWidget implements StatefulWidget {
+            </React.Fragment>;
+        }
+
+        // if (CompositeTreeNode.is(node)) {
+        //     return <React.Fragment>
+        //         {decorationData.map((decoration, index) => {
+        //             const { tooltip } = decoration;
+        //             const { fontData } = decoration as TreeDecoration.TailDecoration;
+        //             const color = (decoration as TreeDecoration.TailDecorationIcon).color;
+        //             const icon = this.getGenericClassIconDecoration();
+        //             const className = [TREE_NODE_SEGMENT_CLASS, TREE_NODE_TAIL_CLASS].join(' ');
+        //             const style = fontData ? this.applyFontStyles({}, fontData) : color ? { color } : undefined;
+        //             const content = <span key={node.id + 'icon' + index} className={this.getIconClass(icon, [TreeDecoration.Styles.DECORATOR_SIZE_CLASS])}></span>;
+        //             return <div key={node.id + className + index} className={className} style={style} title={tooltip}>
+        //                 {content}
+        //             </div>;
+        //         })}
+        //     </React.Fragment>;
+        // }
+        // return <React.Fragment>
+        //     {decorationData.map((decoration, index) => {
+        //         const { tooltip } = decoration;
+        //         const { data, fontData } = decoration as TreeDecoration.TailDecoration;
+        //         const color = (decoration as TreeDecoration.TailDecorationIcon).color;
+        //         const icon = (decoration as TreeDecoration.TailDecorationIcon).icon || (decoration as TreeDecoration.TailDecorationIconClass).iconClass;
+        //         const className = [TREE_NODE_SEGMENT_CLASS, TREE_NODE_TAIL_CLASS].join(' ');
+        //         const style = fontData ? this.applyFontStyles({}, fontData) : color ? { color } : undefined;
+        //         const content = data ? data : icon ? <span key={node.id + 'icon' + index} className={this.getIconClass(icon)}></span> : '';
+        //         return <div key={node.id + className + index} className={className} style={style} title={tooltip}>
+        //             {content}
+        //         </div>;
+        //     })}
+        // </React.Fragment>;
```
We should remove commented code.

vince-fugnitto commented on packages/markers/src/browser/problem/problem-decorator.ts:
```diff
@@ -129,11 +129,19 @@ export class ProblemDecorator implements TreeDecorator {
         const icon = this.getOverlayIcon(marker);
         const color = this.getOverlayIconColor(marker);
         const priority = this.getPriority(marker);
+        //const isResource = Array.from(this.problemManager.getUris()).some(m => m === marker.uri);
+        //const problemStat = this.problemManager.getProblemStat();
         return {
             priority,
             fontData: {
                 color,
             },
+            tailDecorations: [
+                {
+                    data: 'allo',
+                    color
+                }
+            ],
             iconOverlay: {
                 position,
                 icon,
```
We should revert the changes when upstreaming.
Alternatively, we can have a separate commit which adds this data for test purposes (we can drop the commit when it would be ready to merge).

vince-fugnitto commented on packages/scm/src/browser/decorations/scm-navigator-decorator.ts:
```diff
@@ -58,6 +58,7 @@ export class ScmNavigatorDecorator implements TreeDecorator {
                 }
             }
         }
+
```
We can revert this file, as the changes are unnecessary.

vince-fugnitto reviewed (commented):
Please be sure to squash when ready and provide a meaningful commit title and message (it should also be in english).

vince-fugnitto commented on packages/core/src/browser/tree/tree-widget.tsx:
```diff
@@ -729,54 +729,82 @@ export class TreeWidget extends ReactWidget implements StatefulWidget {
+            // If the node is a composite, we just want to use the decorationData with the highest priority (last element).
+            // We put this decorationData in a new array so we can apply the map method on it
+            let arrayHighPrio: (TreeDecoration.TailDecoration | TreeDecoration.TailDecorationIcon | TreeDecoration.TailDecorationIconClass)[] = [];
+            arrayHighPrio.push(decorationData.pop() as TreeDecoration.TailDecoration | TreeDecoration.TailDecorationIcon | TreeDecoration.TailDecorationIconClass);
```
The name of the variable does not mean much as is.
Also, why do we need to perform a map on a single element?

vince-fugnitto reviewed (commented):
Please fix build errors:

```
@theia/core:   758:1   error  asterisks in jsdoc must be aligned (tslint:jsdoc-format)  @typescript-eslint/tslint/config
@theia/core:   776:17  error  'arrayHighPrio' is never reassigned. Use 'const' instead  prefer-const
@theia/core:   791:35  error  Missing semicolon   
```

paul-marechal commented on packages/core/src/browser/tree/tree-widget.tsx:
```diff
@@ -729,54 +729,82 @@ export class TreeWidget extends ReactWidget implements StatefulWidget {
+     */
+    protected getGenericClassIconDecoration(): string[] {
+        return ['fa', 'fa-circle'];
+    }
```
JSDoc looks weird.

vince-fugnitto commented on packages/markers/src/browser/problem/problem-decorator.ts:
```diff
@@ -129,11 +129,19 @@ export class ProblemDecorator implements TreeDecorator {
         const icon = this.getOverlayIcon(marker);
         const color = this.getOverlayIconColor(marker);
         const priority = this.getPriority(marker);
+        //const isResource = Array.from(this.problemManager.getUris()).some(m => m === marker.uri);
+        //const problemStat = this.problemManager.getProblemStat();
```
You can remove the following lines, it fails the build due to a missing whitespace.

```
lerna ERR! /home/evinfug/workspaces/omar-theia/packages/markers/src/browser/problem/problem-decorator.ts
lerna ERR!   132:9  error  Expected exception block, space or tab after '//' in comment  spaced-comment
lerna ERR!   133:9  error  Expected exception block, space or tab after '//' in comment  spaced-comment
```

vince-fugnitto replied:
On a side note, you can enable [github actions](https://github.com/features/actions) on your repository through the settings to enable CI (so you can verify if your pull-requests build and tests pass):

![image](https://user-images.githubusercontent.com/40359487/117998741-2c5ce200-b312-11eb-91f0-cc51d5319b57.png)

paul-marechal commented on packages/core/src/browser/tree/tree-widget.tsx:
```diff
@@ -729,54 +729,75 @@ export class TreeWidget extends ReactWidget implements StatefulWidget {
+
+        if (CompositeTreeNode.is(node)) {
+            // If the node is a composite, we just want to use the decorationData with the highest priority (last element).
+            const highPriorityDecoration = decorationData.pop();
```
This can be undefined, you should handle that case.

vince-fugnitto commented on packages/core/src/browser/tree/tree-widget.tsx:
```diff
@@ -763,14 +768,26 @@ export class TreeWidget extends ReactWidget implements StatefulWidget {
-                const icon = (decoration as TreeDecoration.TailDecorationIcon).icon || (decoration as TreeDecoration.TailDecorationIconClass).iconClass;
                 const className = [TREE_NODE_SEGMENT_CLASS, TREE_NODE_TAIL_CLASS].join(' ');
                 const style = fontData ? this.applyFontStyles({}, fontData) : color ? { color } : undefined;
+                const icon = (decoration as TreeDecoration.TailDecorationIcon).icon || (decoration as TreeDecoration.TailDecorationIconClass).iconClass;
```
minor: we can revert this change, it only moves the variable.

vince-fugnitto commented on packages/core/src/browser/tree/tree-widget.tsx:
```diff
@@ -763,14 +768,26 @@ export class TreeWidget extends ReactWidget implements StatefulWidget {
+        return this.renderTailDecorationsForNode(node, props, tailDecorations);
+    }
+
+    /**
+     * Render the tree node tail decorations if the node is not a `CompositeTreeNode`.
+     */
```
Documentation is misleading, we can remove it entirely.

vince-fugnitto commented on packages/navigator/src/browser/navigator-widget.tsx:
```diff
@@ -139,6 +143,36 @@ export class FileNavigatorWidget extends FileTreeWidget {
+            return;
+        }
+
+        if (CompositeTreeNode.is(node)) {
+            return this.renderTailDecorationsForCompositeNode(node, props, tailDecorations);
+        } else {
+            return this.renderTailDecorationsForNode(node, props, tailDecorations);
+        }
+    }
```
We can make use of `FileStatNode` and rename the composite method:

```suggestion
        if (FileStatNode.is(node) && node.fileState.isDirectory) {
            return this.renderTailDecorationsForDirectoryNode(node, props, tailDecorations);
        } else {
            return this.renderTailDecorationsForNode(node, props, tailDecorations);
        }
    }
```

vince-fugnitto commented on packages/navigator/src/browser/navigator-widget.tsx:
```diff
@@ -139,6 +143,36 @@ export class FileNavigatorWidget extends FileTreeWidget {
+
+    protected renderTailDecorationsForCompositeNode(node: TreeNode, props: NodeProps, tailDecorations:
+        (TreeDecoration.TailDecoration | TreeDecoration.TailDecorationIcon | TreeDecoration.TailDecorationIconClass)[]): React.ReactNode {
+        // If the node is a composite, we just want to use the decorationData with the highest priority (last element).
```
```suggestion
        // If the node represents a directory, we just want to use the decorationData with the highest priority (last element).
```

OmarSdt-EC committed:
tree: use dot decoration on folders
The following commit updates the rendering of tail decorations for
composite tree nodes (nodes that contain children nodes like folders).

The following updates the rendering of tail decorations for composite
tree nodes (nodes that contain children nodes like folders).  The
updates include :
- rendering a generic icon decoration (dot) for composite nodes with children with decoration data.
- updates to the rendering logic to only render the decoration data with the highest priority so no duplicate generic icons are present.
